fixes #153 add project/relatedProjects array

pindec opened the pull request:
fixes #153

pindec committed:
fixes #153 add project/relatedProjects array

duncandewhurst commented on schema/project-level/codelists/relatedProject.csv:
```diff
@@ -0,0 +1,5 @@
+Code,Title,Description
+construction,Construction,The related project is the initial construction of the infrastructure
```
```suggestion
construction,Construction,The related project is the initial construction of the set of infrastructure assets
```

duncandewhurst commented on schema/project-level/codelists/relatedProject.csv:
```diff
@@ -0,0 +1,5 @@
+Code,Title,Description
+construction,Construction,The related project is the initial construction of the infrastructure
+rehabilitation,Rehabilitation,The related project might result in the rehabilitation of the same infrastructure
```
```suggestion
rehabilitation,Rehabilitation,The related project might result in the rehabilitation of the same set of infrastructure assets
```

duncandewhurst commented on schema/project-level/codelists/relatedProject.csv:
```diff
@@ -0,0 +1,5 @@
+Code,Title,Description
+construction,Construction,The related project is the initial construction of the infrastructure
+rehabilitation,Rehabilitation,The related project might result in the rehabilitation of the same infrastructure
+replacement,Replacement,The related project might result in the replacement of the same infrastructure
```
```suggestion
replacement,Replacement,The related project might result in the replacement of the same set of infrastructure assets
```

duncandewhurst commented on schema/project-level/codelists/relatedProject.csv:
```diff
@@ -0,0 +1,5 @@
+construction,Construction,The related project is the initial construction of the infrastructure
+rehabilitation,Rehabilitation,The related project might result in the rehabilitation of the same infrastructure
+replacement,Replacement,The related project might result in the replacement of the same infrastructure
+expansion,Expansion,The related project might result in the expansion of the same infrastructure
```
```suggestion
expansion,Expansion,The related project might result in the expansion of the same set of infrastructure assets
```

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -85,6 +85,14 @@
     },
+    "relatedProjects": {
+      "title": "Related projects",
+      "description": "The details of projects related by being involved with the same infrastructure as this project. For example, a related project represented under a separate Open Contracting for Infrastructure Data Standard project identifier (OC4IDS id).",
```
Align language with definition of `project` and make example real-world:

```suggestion
      "description": "The details of projects related to the same set of infrastructure assets as this project. For example, to link a project for the replacement of a bridge to the project for its initial construction.",
```

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1358,6 +1366,66 @@
+          "description": "The type of relationship, using the open [relatedProjects](https://standard.open-contracting.org/infrastructure/{{version}}/{{lang}}/reference/codelists/#relatedproject) codelist.",
+          "type": [
+            "array",
+            "null"
```
Remove `null` as this not required in OC4IDS because there is no release model or merging and make `type` a scalar value, i.e. `"type": "array"`

Please remove `null` for the other properties too.

Edit: in fact, since the `relatedProcess` codelist matches the `projectType` codelist and `project/type` is not an array, I think we should only allow one value here (i.e. make it `string` not `array`)

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1358,6 +1366,66 @@
+      "fdescription": "A reference to a project related to this project. Generally related projects either precede or follow on from the current project.",
+      "type": "object",
+      "properties": {
+        "id": {
```
Make `id` required:

1. Add a `required` array to the `RelatedProject` object and add `id` to the array
2. Add `"minLength": 1` to the `id` object

See the `Modification` definition for an example.

See [schema conventions - id fields](https://ocds-standard-development-handbook.readthedocs.io/en/latest/standard/conventions.html#id-fields) in the standard development handbook for why.

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1358,6 +1366,66 @@
+          ],
+          "codelist": "relatedProject.csv",
+          "openCodelist": true,
+          "items": {
```
See above comment, once `"type": "string"` this can be removed.

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1358,6 +1366,66 @@
+        },
+        "identifier": {
+          "title": "Identifier",
+          "description": "The identifier of the related project. If the scheme is 'oc4ids', this must be an Open Contracting for Infrastructure Data Standard project ID.",
```
Just for brevity:

```suggestion
          "description": "The identifier of the related project. If the scheme is 'oc4ids', this must be an OC4IDS project ID.",
```

duncandewhurst commented on schema/project-level/codelists/relatedProjectScheme.csv:
```diff
@@ -0,0 +1,2 @@
+Code,Title,Description
+oc4ids,Open Contracting for Infrastructure Data Standard project identifier,An OC4IDS project id
```
```suggestion
oc4ids,OC4IDS project id,An Open Contracting for Infrastructure Data Standard project identifier
```

duncandewhurst commented on docs/reference/schema.md:
```diff
@@ -218,6 +218,19 @@ Where the source system does not contain time information, a judgement should be
 
+### RelatedProject
+
+Details of a related project. A project is related when it is involved with the same infrastructure as the current project.
```
```suggestion
Details of a project related to the same set of infrastructure assets as the current project.
```

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1358,6 +1366,66 @@
+        },
+        "scheme": {
+          "title": "Scheme",
+          "description": "The identification scheme used by this cross-reference, using the open [relatedProjectScheme]((https://standard.open-contracting.org/infrastructure/{{version}}/{{lang}}/reference/codelists/#relatedprojectscheme) codelist.",
```
```suggestion
          "description": "The identification scheme used by this cross-reference, using the open [relatedProjectScheme](https://standard.open-contracting.org/infrastructure/{{version}}/{{lang}}/reference/codelists/#relatedprojectscheme) codelist.",
```

duncandewhurst reviewed (changes requested):
Thanks Charlie! Looking really good, just a few things to address. Mainly aligning language with the definition of project in OC4IDS and removing some unneeded bits from the schema.

Please also add `relatedProjects` to the worked example JSON file and implementation guidance page.

pindec committed:
Update schema/project-level/codelists/relatedProject.csv

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/codelists/relatedProject.csv

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/codelists/relatedProject.csv

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/codelists/relatedProject.csv

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/project-schema.json

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/project-schema.json

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/project-schema.json

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update docs/reference/schema.md

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Update schema/project-level/codelists/relatedProjectScheme.csv

Co-Authored-By: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

pindec committed:
Updating example, example guidance and schema for related projects

pindec commented:
In the example, type is now "expansion" and linked to relatedProject for the original construction.

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1369,63 +1369,50 @@
-          "type": "string"
+          "type": [
+            "string",
+            "integer"
```
Remove 'integer' from `type`, per https://github.com/open-contracting/infrastructure/issues/184

(sorry - I know it was included in the example I suggested!)

duncandewhurst reviewed (approved):
Happy for this to be merged once the final comment is addressed.

jpmckinney committed:
Review addition of relatedProjects

- Remove integer type for RelatedProject.id
- Quote id value and fix indentation in example.json
- Add periods to code descriptions in relatedProject.csv
- Use sentence, instead of sentence fragment
- Fix typos and punctuation
- Edit for clarity

jpmckinney committed:
Re-order fields in RelatedProject to put most important first

jpmckinney committed:
Merge branch '0.9-dev' into 153_add_relatedProjects_array

jpmckinney commented:
I fixed the comment (removing integer type) and made small edits. The most substantive edit is referring to the related projects data as "references" rather than "details" about related projects, since it contains no details.

jpmckinney commented:
@duncandewhurst @pindec Feel free to merge if my first commit looks good (my second commit just reorders the new fields, and my last commit fixes a conflict in the changelog).

jpmckinney committed:
Merge branch '0.9-dev' into 153_add_relatedProjects_array

jpmckinney committed:
Merge branch '0.9-dev' into 153_add_relatedProjects_array
